File: src/card.py
import pygame

# Colors
WHITE = (255, 255, 255)

# Visual mappings for drawing (category values -> visual representation)
COLOR_MAP = {
    "red": (220, 60, 60),
    "green": (60, 180, 60),
    "blue": (60, 100, 220),
}


class Card:
    def __init__(self, category1, category2, category3, category4):
        self.category1 = category1  # shape
        self.category2 = category2  # color
        self.category3 = category3  # count
        self.category4 = category4  # fill

    def __eq__(self, other):
        return (self.category1 == other.category1 and self.category2 == other.category2 and
                self.category3 == other.category3 and self.category4 == other.category4)

    def check_if_set(self, other1, other2):
        def check_attribute(attr, other1_attr, other2_attr):
            all_same = (attr == other1_attr == other2_attr)
            all_diff = (attr != other1_attr and attr != other2_attr and other1_attr != other2_attr)
            return all_same or all_diff

        attrs = [self.category1, self.category2, self.category3, self.category4]
        other1_attrs = [other1.category1, other1.category2, other1.category3, other1.category4]
        other2_attrs = [other2.category1, other2.category2, other2.category3, other2.category4]
        return all(check_attribute(a, b, c) for a, b, c in zip(attrs, other1_attrs, other2_attrs))

    def draw(self, screen, x, y, width, height):
        """Draw this card's shapes at the given position."""
        color = COLOR_MAP.get(self.category2, WHITE)
        count = int(self.category3)
        shape = self.category1
        fill = self.category4

        shape_width = 20
        shape_height = 30
        spacing = 24
        total_width = count * shape_width + (count - 1) * (spacing - shape_width)
        start_x = x + (width - total_width) // 2
        center_y = y + height // 2

        for i in range(count):
            sx = start_x + i * spacing
            sy = center_y - shape_height // 2
            self._draw_shape(screen, shape, fill, sx, sy, shape_width, shape_height, color)

    def _draw_shape(self, screen, shape, fill, x, y, w, h, color):
        """Draw a single shape with the given fill style."""
        if shape == "circle":
            rect = pygame.Rect(x, y, w, h)
            self._draw_filled(screen, "ellipse", rect, fill, color)
        elif shape == "square":
            rect = pygame.Rect(x + 2, y + 4, w - 4, h - 8)
            self._draw_filled(screen, "rect", rect, fill, color)
        elif shape == "triangle":
            points = [(x + w // 2, y), (x + w, y + h), (x, y + h)]
            self._draw_filled(screen, "polygon", points, fill, color)

    def _draw_filled(self, screen, shape_type, shape_data, fill, color):
        """Draw shape with the specified fill style."""
        if fill == "solid":
            if shape_type == "polygon":
                pygame.draw.polygon(screen, color, shape_data)
            elif shape_type == "ellipse":
                pygame.draw.ellipse(screen, color, shape_data)
            elif shape_type == "rect":
                pygame.draw.rect(screen, color, shape_data)
        elif fill == "open":
            if shape_type == "polygon":
                pygame.draw.polygon(screen, color, shape_data, 2)
            elif shape_type == "ellipse":
                pygame.draw.ellipse(screen, color, shape_data, 2)
            elif shape_type == "rect":
                pygame.draw.rect(screen, color, shape_data, 2)
        elif fill == "striped":
            # Draw outline + horizontal stripes
            if shape_type == "polygon":
                pygame.draw.polygon(screen, color, shape_data, 2)
                min_y = min(p[1] for p in shape_data)
                max_y = max(p[1] for p in shape_data)
                min_x = min(p[0] for p in shape_data)
                max_x = max(p[0] for p in shape_data)
            elif shape_type == "ellipse":
                pygame.draw.ellipse(screen, color, shape_data, 2)
                min_y, max_y = shape_data.top, shape_data.bottom
                min_x, max_x = shape_data.left, shape_data.right
            elif shape_type == "rect":
                pygame.draw.rect(screen, color, shape_data, 2)
                min_y, max_y = shape_data.top, shape_data.bottom
                min_x, max_x = shape_data.left, shape_data.right
            for stripe_y in range(int(min_y) + 4, int(max_y), 6):
                pygame.draw.line(screen, color, (min_x + 2, stripe_y), (max_x - 2, stripe_y), 1)

File: src/card.ts
import type { Shape, Color, Count, Fill } from './types';
import { getShapeCanvas, SHAPE_SIZE } from './shapes';
import { SHAPE_SPACING } from './constants';

export class Card {
  readonly shape: Shape;
  readonly color: Color;
  readonly count: Count;
  readonly fill: Fill;

  constructor(shape: Shape, color: Color, count: Count, fill: Fill) {
    this.shape = shape;
    this.color = color;
    this.count = count;
    this.fill = fill;
  }

  checkIfSet(other1: Card, other2: Card): boolean {
    const checkAttr = <T>(a: T, b: T, c: T): boolean => {
      const allSame = a === b && b === c;
      const allDiff = a !== b && a !== c && b !== c;
      return allSame || allDiff;
    };

    return (
      checkAttr(this.shape, other1.shape, other2.shape) &&
      checkAttr(this.color, other1.color, other2.color) &&
      checkAttr(this.count, other1.count, other2.count) &&
      checkAttr(this.fill, other1.fill, other2.fill)
    );
  }

  draw(ctx: CanvasRenderingContext2D, x: number, y: number, width: number, height: number): void {
    const count = parseInt(this.count, 10);
    const shapeCanvas = getShapeCanvas(this.shape, this.color, this.fill);
    if (!shapeCanvas) return;

    const totalWidth = count * SHAPE_SIZE + (count - 1) * (SHAPE_SPACING - SHAPE_SIZE);
    const startX = x + Math.floor((width - totalWidth) / 2);
    const centerY = y + Math.floor((height - SHAPE_SIZE) / 2);

    // drawImage is faster than putImageData because it uses GPU acceleration
    for (let i = 0; i < count; i++) {
      ctx.drawImage(shapeCanvas, startX + i * SHAPE_SPACING, centerY);
    }
  }

  equals(other: Card): boolean {
    return this.shape === other.shape && this.color === other.color &&
           this.count === other.count && this.fill === other.fill;
  }

  toString(): string {
    return `Card(${this.shape}, ${this.color}, ${this.count}, ${this.fill})`;
  }
}

export function createDeck(): Card[] {
  const deck: Card[] = [];
  const shapes: Shape[] = ['diamond', 'oval', 'squiggle'];
  const colors: Color[] = ['red', 'green', 'purple'];
  const counts: Count[] = ['1', '2', '3'];
  const fills: Fill[] = ['solid', 'striped', 'empty'];

  for (const shape of shapes) {
    for (const color of colors) {
      for (const count of counts) {
        for (const fill of fills) {
          deck.push(new Card(shape, color, count, fill));
        }
      }
    }
  }
  return deck;
}

export function shuffleDeck(deck: Card[]): void {
  for (let i = deck.length - 1; i > 0; i--) {
    const j = Math.floor(Math.random() * (i + 1));
    [deck[i], deck[j]] = [deck[j], deck[i]];
  }
}

export function hasValidSet(cards: (Card | null)[]): boolean {
  const valid = cards.filter((c): c is Card => c !== null);
  if (valid.length < 3) return false;

  for (let i = 0; i < valid.length - 2; i++) {
    for (let j = i + 1; j < valid.length - 1; j++) {
      for (let k = j + 1; k < valid.length; k++) {
        if (valid[i].checkIfSet(valid[j], valid[k])) return true;
      }
    }
  }
  return false;
}

export function findAllValidSets(cards: (Card | null)[]): [number, number, number][] {
  const validSets: [number, number, number][] = [];
  for (let i = 0; i < cards.length - 2; i++) {
    if (!cards[i]) continue;
    for (let j = i + 1; j < cards.length - 1; j++) {
      if (!cards[j]) continue;
      for (let k = j + 1; k < cards.length; k++) {
        if (!cards[k]) continue;
        if (cards[i]!.checkIfSet(cards[j]!, cards[k]!)) {
          validSets.push([i, j, k]);
        }
      }
    }
  }
  return validSets;
}
